### PR TITLE
Python interface for updating translation and rotation rates of markers

### PR DESCRIPTION
**Pull request by HahsFilip**
## Proposed Changes
*Give a brief overview of your contribution here in a few sentences.*
 My changes allow the changing of moving wall velocities when using the py_wrapper. 

## Related Work
*Resolve any issues (bug fix or feature request), note any related PRs, or mention interactions with the work of others, if any.*
There has been some previous work, where functionality for changing mesh rotation rates was added.

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [X ] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with --warnlevel=3 when using meson).
- [x] My contribution is commented and consistent with SU2 style (https://su2code.github.io/docs_v7/Style-Guide/).
- [x] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp), if necessary.

### CODE REVIEW

**Comment by HahsFilip** [comment]
@pcarruscag This is my first PR, so I hope that at least a few things will be ok :)

File: TestCases/py_wrapper/rotating_cylinder/run.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import sys
+from optparse import OptionParser	# use a parser for configuration
+import pysu2			            # imports the SU2 wrapped module
+from math import *
+import os
+import pysu2ad
+from mpi4py import MPI
+import numpy as np
+import shutil
```
[the remark this answers is not in the document]
Comment (HahsFilip) [reply TestCases/py_wrapper/rotating_cylinder/run.py]: Resolves

File: TestCases/py_wrapper/rotating_cylinder/run.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import sys
+from optparse import OptionParser	# use a parser for configuration
+import pysu2			            # imports the SU2 wrapped module
+from math import *
+import os
+import pysu2ad
```
[the remark this answers is not in the document]
Comment (HahsFilip) [reply TestCases/py_wrapper/rotating_cylinder/run.py]: Resolved

File: TestCases/py_wrapper/rotating_cylinder/run.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import sys
+from optparse import OptionParser	# use a parser for configuration
+import pysu2			            # imports the SU2 wrapped module
+from math import *
+import os
```
[the remark this answers is not in the document]
Comment (HahsFilip) [reply TestCases/py_wrapper/rotating_cylinder/run.py]: Resolved

File: TestCases/py_wrapper/rotating_cylinder/run.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import sys
+from optparse import OptionParser	# use a parser for configuration
```
[the remark this answers is not in the document]
Comment (HahsFilip) [reply TestCases/py_wrapper/rotating_cylinder/run.py]: Resolved

File: TestCases/py_wrapper/rotating_cylinder/run.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import sys
```
[the remark this answers is not in the document]
Comment (HahsFilip) [reply TestCases/py_wrapper/rotating_cylinder/run.py]: Resolved

File: Common/include/CConfig.hpp [hunk 6]
```diff
@@ -5962,6 +5971,16 @@ class CConfig {
    */
   su2double GetMarkerRotationRate(unsigned short iMarkerMoving, unsigned short iDim) const { return MarkerRotation_Rate[3*iMarkerMoving + iDim];}
 
+  /*!
+   * \brief Set the rotation rate of the marker.
+   * \param[in] iMarkerMoving -  Index of the moving marker (as specified in Marker_Moving)
+   * \param[in] iDim - spatial component
+   * \param[in] val - Rotational velocity
+   */
+  void SetMarkerRotation_Rate(unsigned short iMarkerMoving, unsigned short iDim, su2double val) {
```
Comment (pcarruscag) [inline Common/include/CConfig.hpp:5980]:
```suggestion
  void SetMarkerRotationRate(unsigned short iMarkerMoving, unsigned short iDim, su2double val) {
```

File: SU2_CFD/src/python_wrapper_structure.cpp [hunk 7]
```diff
@@ -167,3 +167,19 @@ void CDriver::SetRotationRate(passivedouble rot_x, passivedouble rot_y, passived
...
+    config_container[iZone]->SetMarkerRotation_Rate(iMarker, 1, rot_y);
+    config_container[iZone]->SetMarkerRotation_Rate(iMarker, 2, rot_z);
+  }
+}
+
+void CDriver::SetMarkerTranslationRate(unsigned short iMarker, passivedouble vel_x, passivedouble vel_y, passivedouble vel_z) {
+  for (iZone = 0; iZone < nZone; iZone++) {
+    config_container[iZone]->SetMarkerTranslation_Rate(iMarker, 0, vel_x);
+    config_container[iZone]->SetMarkerTranslation_Rate(iMarker, 1, vel_y);
+    config_container[iZone]->SetMarkerTranslation_Rate(iMarker, 2, vel_z);
```
Comment (pcarruscag) [inline SU2_CFD/src/python_wrapper_structure.cpp:182]:
```suggestion
    config_container[iZone]->SetMarkerRotationRate(iMarker, 0, rot_x);
    config_container[iZone]->SetMarkerRotationRate(iMarker, 1, rot_y);
    config_container[iZone]->SetMarkerRotationRate(iMarker, 2, rot_z);
  }
}

void CDriver::SetMarkerTranslationRate(unsigned short iMarker, passivedouble vel_x, passivedouble vel_y, passivedouble vel_z) {
  for (iZone = 0; iZone < nZone; iZone++) {
    config_container[iZone]->SetMarkerTranslationRate(iMarker, 0, vel_x);
    config_container[iZone]->SetMarkerTranslationRate(iMarker, 1, vel_y);
    config_container[iZone]->SetMarkerTranslationRate(iMarker, 2, vel_z);
```

File: TestCases/py_wrapper/rotating_cylinder/run.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
+import pysu2			            # imports the SU2 wrapped module
```
Comment (pcarruscag) [inline TestCases/py_wrapper/rotating_cylinder/run.py:1]: Thanks :+1: to make a regression test from this you need to add it to parallel_regression.py

File: TestCases/py_wrapper/rotating_cylinder/run.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+from math import *
+from mpi4py import MPI
+import numpy as np
+
+comm = MPI.COMM_WORLD
+rank = comm.Get_rank()
+ad_com = MPI.COMM_WORLD
+rank_ad = ad_com.Get_rank()
```
Comment (kursatyurt) [inline TestCases/py_wrapper/rotating_cylinder/run.py:9]:
```suggestion
```

Unused lines

File: TestCases/py_wrapper/rotating_cylinder/run.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import pysu2			            # imports the SU2 wrapped module
+from math import *
```
Comment (kursatyurt) [inline TestCases/py_wrapper/rotating_cylinder/run.py:2]:
```suggestion
```

This is not used as well.

**Comment by HahsFilip** [comment]
I'm having some trouble getting the regression tests to pass - for some reason there is a result difference in the moving wall tests. I'm not sure how this can be possible, considering I am not changing any values.  The convergence seems to be slowed down considerably, but I'm having trouble identifying which function may be causing this.

**Comment by pcarruscag** [comment]
Why do you say the convergence is slowed down? It looks like a very small change. Did you run the cases for more iterations?

**Comment by HahsFilip** [comment]
Well, I've noticed the problem, because the unit tests don't pass - cavity and spinning cylinder especially. They exceed the number of iterations, and at the same time there is a result discrepancy.

**Comment by HahsFilip** [comment]
I'm actually having the same problems running the code on the develop branch....

**Comment by HahsFilip** [comment]
So you mean I should run the tests to convergence, and compare the results between branches? I wonder why compilation problems could be the culprit, as the tests fail both locally and in the github test. Shouldn't the github test always be compiled the same way?

**Comment by pcarruscag** [comment]
You mentioned issues with the develop branch, those are likely due to compilation differences.
Yes run to convergence with both branches and compare please. I expect integrated quantities will be exactly the same.

File: SU2_CFD/include/drivers/CDriver.hpp [hunk 11]
```diff
@@ -535,7 +535,24 @@ class CDriver : public CDriverBase {
    * \param[in] rot_z - Value of Angular velocity about z-axes.
    */
   void SetRotationRate(passivedouble rot_x, passivedouble rot_y, passivedouble rot_z);
+  /*!
```
Comment (pcarruscag) [inline SU2_CFD/include/drivers/CDriver.hpp:538]:
```suggestion

  /*!
```

File: SU2_CFD/src/python_wrapper_structure.cpp [hunk 12]
```diff
@@ -167,3 +167,19 @@ void CDriver::SetRotationRate(passivedouble rot_x, passivedouble rot_y, passived
...
+    config_container[iZone]->SetMarkerRotationRate(iMarker, 2, rot_z);
+  }
+}
+
+void CDriver::SetMarkerTranslationRate(unsigned short iMarker, passivedouble vel_x, passivedouble vel_y, passivedouble vel_z) {
+  for (iZone = 0; iZone < nZone; iZone++) {
+    config_container[iZone]->SetMarkerTranslationRate(iMarker, 0, vel_x);
+    config_container[iZone]->SetMarkerTranslationRate(iMarker, 1, vel_y);
+    config_container[iZone]->SetMarkerTranslationRate(iMarker, 2, vel_z);
+  }
```
Comment (pcarruscag) [inline SU2_CFD/src/python_wrapper_structure.cpp:183]:
```suggestion
  config_container[selected_zone]->SetMarkerRotationRate(iMarker, 0, rot_x);
  config_container[selected_zone]->SetMarkerRotationRate(iMarker, 1, rot_y);
  config_container[selected_zone]->SetMarkerRotationRate(iMarker, 2, rot_z);
}

void CDriver::SetMarkerTranslationRate(unsigned short iMarker, passivedouble vel_x, passivedouble vel_y, passivedouble vel_z) {
  config_container[selected_zone]->SetMarkerTranslationRate(iMarker, 0, vel_x);
  config_container[selected_zone]->SetMarkerTranslationRate(iMarker, 1, vel_y);
  config_container[selected_zone]->SetMarkerTranslationRate(iMarker, 2, vel_z);
```

**Review by pcarruscag — COMMENTED** [review]
Hey @HahsFilip do you think you can wrap this up?
Take care of Kursat's comments and update the failing tests since it seem to be just numerical noise (or do you have reason to believe otherwise?)

**Comment by pcarruscag** [comment]
You also need to add yourself to the AUTHORS.md file

**Comment by HahsFilip** [comment]
@pcarruscag, sorry for the delay I had to get my life together a bit after the conference :). I think I have the correct changes, I have changed the tests and removed unused lines.

Filip

File: TestCases/py_wrapper/rotating_cylinder/run.py (hunk 9 above)
Comment (HahsFilip) [reply TestCases/py_wrapper/rotating_cylinder/run.py]: Removed

File: TestCases/py_wrapper/rotating_cylinder/run.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+from mpi4py import MPI
+import numpy as np
+import shutil
+comm = MPI.COMM_WORLD
+rank = comm.Get_rank()
+ad_com = MPI.COMM_WORLD
+rank_ad = ad_com.Get_rank() 
+n_of_steps = 10
+rotation_vector = np.zeros(n_of_steps)+20
+rotation_vector[5:9] = -rotation_vector[5:9]
```
[the remark this answers is not in the document]
Comment (HahsFilip) [reply TestCases/py_wrapper/rotating_cylinder/run.py]: Done

File: TestCases/py_wrapper/rotating_cylinder/run.py (hunk 8 above)
Comment (pcarruscag) [reply TestCases/py_wrapper/rotating_cylinder/run.py:1]: Add our license header to this file please (since it's also code).

**Review by pcarruscag — APPROVED** [review]
LGTM

File: externals/codi
Comment (pcarruscag) [inline externals/codi:1]: You need to revert these changes to submodules, this may be why you see some failing cases

**Comment by kursatyurt** [comment]
@pcarruscag , @HahsFilip 

Do the hybrid tests need an update? Do any of you have any idea about why they are failing?

**Review by pcarruscag — COMMENTED** [review]
Because this branch has the wrong versions of various submodules

**Comment by kursatyurt** [comment]
> Because this branch has the wrong versions of various submodules

I  see I assumed the last commit reverted all. As it is on a fork we cannot revert 👍🏼

**Comment by kursatyurt** [comment]
@pcarruscag, is it possible to merge this PR to some temporary branch in the SU2 repo, fix the broken submodules there, and merge it to the main branch? It looks like @HahsFilip is busy. PR is almost complete but cannot be merged. I will do the required fixes.

**Comment by pcarruscag** [comment]
Thanks @kursatyurt submodules are fixed here https://github.com/su2code/SU2/commit/8b89e857ff72782b0caccc3f63385c03bc67f6f7
If you can do the other fixes and hook up the regression test it would be great

**Comment by kursatyurt** [comment]
Thanks @pcarruscag , it looks like standard regression tests are passed without any problem, and ARM tests are problematic, as always.

**Pull request merged**